hydra-explorer/test/Hydra/Explorer/IntegrationSpec.hs

```diff
@@ -0,0 +1,232 @@
+{-# LANGUAGE DeriveAnyClass #-}
+
+-- | Integration tests for the 'hydra-explorer' executable. These will run
+-- also 'hydra-node' on a devnet and assert correct observation.
+module Hydra.Explorer.IntegrationSpec where
+
+import Hydra.Prelude hiding (get)
+import Test.Hydra.Prelude
+
+-- XXX: Depends on hydra-node for this
+import Hydra.Logging (showLogsOnFailure)
+
+import CardanoClient (RunningNode (..), queryTip)
+import CardanoNode (NodeLog, withCardanoNodeDevnet)
+import Control.Lens ((^.), (^?))
+import Data.Aeson as Aeson
+import Data.Aeson.Lens (key, nth, _Array, _Number, _String)
+import Hydra.Cardano.Api (NetworkId (..), NetworkMagic (..), toNetworkMagic, unFile)
+import Hydra.Cluster.Faucet (FaucetLog, publishHydraScriptsAs, seedFromFaucet_)
+import Hydra.Cluster.Fixture (Actor (..), aliceSk, bobSk, cperiod)
+import Hydra.Cluster.Scenarios (EndToEndLog, singlePartyHeadFullLifeCycle)
+import Hydra.Cluster.Util (chainConfigFor, keysFor)
+import Hydra.Tx.DepositDeadline (DepositDeadline (..))
+import HydraNode (HydraNodeLog, input, send, waitMatch, withHydraNode)
+import Network.HTTP.Simple (getResponseBody, httpJSON, parseRequestThrow)
+import Network.Socket (PortNumber)
+import System.Process.Typed (Process, ProcessConfig, createPipe, getStderr, proc, setStderr, unsafeProcessHandle, withProcessTerm)
+import Test.Network.Ports (withFreePort)
+
+spec :: Spec
+spec = do
+  -- Simple end-to-end scenario with a cardano devnet and head opened/closed by
+  -- a hydra-node, which a hydra-chain-observer reports into a single
+  -- hydra-explorer.
+  it "aggregates observations" $
+    failAfter 60 $
+      showLogsOnFailure "IntegrationSpec" $ \tracer -> do
+        withTempDir "hydra-explorer-history" $ \tmpDir -> do
+          withHydraExplorer $ \explorer -> do
+            withCardanoNodeDevnet (contramap FromCardanoNode tracer) tmpDir $ \node -> do
+              hydraScriptsTxId <- publishHydraScriptsAs node Faucet
+              withChainObserver node explorer $ do
+                -- Open and close a head
+                singlePartyHeadFullLifeCycle (contramap FromScenario tracer) tmpDir node hydraScriptsTxId
+                -- Query client api
+                allHeads <- getHeads explorer
+                length (allHeads ^. _Array) `shouldBe` 1
+                allHeads ^? nth 0 . key "network" . _String `shouldBe` Just "Testnet"
+                allHeads ^? nth 0 . key "networkMagic" . _String `shouldBe` Just (show . toNetworkMagic $ networkId node)
+                -- NOTE: This deliberately pins the latest version of hydra we test against.
+                allHeads ^? nth 0 . key "version" . _String `shouldBe` Just "0.19.0"
+
+  -- TODO: simplify scenarios! We are only interested in the end-to-end
+  -- integration and not whether the hydra-chain-observer really works (that is
+  -- to be covered in the main hydra repo)
```

Comment:
Should I do this? i.e. reduce fidelity, but also complexity of these tests

Reply:
> We should not check that the chain observer works.

Did this now by re-using a `singlePartyOpenAHead` scenario function and dropping one test case. It's more lean now.

> This resource dependency seems wrong to me. withChainObserver :: should not take an Explorer instance solely for the port? It should be the other way around?

Why not? Note that the `Explorer` here is merely a handle on how the explorer is run in the integration test. As we need to configure the `hydra-chain-observer` process to point to an explorer, passing in the knowledge we have about the explorer seems to be the right way around because the chain observer does need to know where the explorer runs.